docs/text/feature_extraction_settings.rst
Feature extraction settings
===========================

When starting a new data science project involving time series you probably want to start by extracting a
comprehensive set of features. Later you can identify which features are relevant for the task at hand.
In the final stages, you probably want to fine tune the parameter of the features to fine tune your models.

You can do all those things with tsfresh. So, you need to know how to control which features are calculated by tsfresh
and how one can adjust the parameters. In this section, we will clarify this.

For the lazy: Just let me calculate some features!
--------------------------------------------------

To calculate a comprehensive set of features, call the :func:`tsfresh.extract_features` method without
passing a ``default_fc_parameters`` or ``kind_to_fc_parameters`` object. This way you will be using the default options,
which will use all the feature calculators in this package, that we consider are OK to return by default.

For the advanced: How do I set the parameters for all kind of time series?
----------------------------------------------------------------------------

After digging deeper into your data, you maybe want to calculate more of a certain type of feature and less of another
type. So, you need to use custom settings for the feature extractors. To do that with tsfresh you will have to use a
custom settings object:

>>> from tsfresh.feature_extraction import ComprehensiveFCParameters
>>> settings = ComprehensiveFCParameters()
>>> # Set here the options of the settings object as shown in the paragraphs below
>>> # ...
>>> from tsfresh.feature_extraction import extract_features
>>> extract_features(df, default_fc_parameters=settings)

The ``default_fc_parameters`` is expected to be a dictionary which maps feature calculator names
(the function names you can find in the :mod:`tsfresh.feature_extraction.feature_calculators` file) to a list
of dictionaries, which are the parameters with which the function will be called (as key value pairs). Each
function-parameter combination that is in this dict will be called during the extraction and will produce a feature.
If the function does not take any parameters, the value should be set to ``None``.

For example:

.. code:: python

    fc_parameters = {
        "length": None,
        "large_standard_deviation": [{"r": 0.05}, {"r": 0.1}]
    }

will produce three features: one by calling the
:func:`tsfresh.feature_extraction.feature_calculators.length` function without any parameters and two by calling
:func:`tsfresh.feature_extraction.feature_calculators.large_standard_deviation` with ``r = 0.05`` and ``r = 0.1``.

So you can control which features will be extracted, by adding or removing either keys or parameters from this dict.
It is as easy as that.
If you decide not to calculate the length feature here, you delete it from the dictionary:

.. code:: python

    del fc_parameters["length"]

And now, only the two other features are calculated.

For convenience, three dictionaries are predefined and can be used right away:

* :class:`tsfresh.feature_extraction.settings.ComprehensiveFCParameters`: includes all features without parameters and
  all features with parameters, each with different parameter combinations. This is the default for :func:`tsfresh.extract_features`
  if you do not hand in a ``default_fc_parameters`` at all.
* :class:`tsfresh.feature_extraction.settings.MinimalFCParameters`: includes only a handful of features
  and can be used for quick tests. The features which have the "minimal" attribute are used here.
* :class:`tsfresh.feature_extraction.settings.EfficientFCParameters`: Mostly the same features as in the
  :class:`tsfresh.feature_extraction.settings.ComprehensiveFCParameters`, but without features which are marked with the
  "high_comp_cost" attribute. This can be used if runtime performance plays a major role.

Theoretically, you could calculate an unlimited number of features with tsfresh by adding entry after entry to the
dictionary.


For the ambitious: How do I set the parameters for different type of time series?
---------------------------------------------------------------------------------

It is also possible to control the features to be extracted for the different kinds of time series individually.
You can do so by passing a ``kind_to_fc_parameters`` parameter to the :func:`tsfresh.extract_features` function.
It should be a dict mapping from kind names (as string) to ``fc_parameters`` objects,
which you would normally pass as an argument to the ``default_fc_parameters`` parameter.

So, for example the following code snippet:

.. code:: python

    kind_to_fc_parameters = {
        "temperature": {"mean": None},
        "pressure": {"maximum": None, "minimum": None}
    }

will extract the ``"mean"`` feature of the ``"temperature"`` time series and the ``"minimum"`` and ``"maximum"`` of the
``"pressure"`` time series.

The ``kind_to_fc_parameters`` argument will partly override the ``default_fc_parameters``. So, if you include a kind
name in the ``kind_to_fc_parameters`` parameter, its value will be used for that kind.
Other kinds will still use the ``default_fc_parameters``.


A handy trick: Do I really have to create the dictionary by hand?
-----------------------------------------------------------------

Not necessarily. Let's assume you have a DataFrame of tsfresh features.
By using feature selection algorithms you find out that only a subgroup of features is relevant.


Then, we provide the :func:`tsfresh.feature_extraction.settings.from_columns` method that constructs the ``kind_to_fc_parameters``
dictionary from the column names of this filtered feature matrix to make sure that only relevant features are extracted.

This can save a huge amount of time because you prevent the calculation of unnecessary features.
Let's illustrate this with an example:

.. code:: python

    # X_tsfresh contains the extracted tsfresh features
    X_tsfresh = extract_features(...)

    # which are now filtered to only contain relevant features
    X_tsfresh_filtered = some_feature_selection(X_tsfresh, y, ....)

    # we can easily construct the corresponding settings object
    kind_to_fc_parameters = tsfresh.feature_extraction.settings.from_columns(X_tsfresh_filtered)

The above code will construct for you the ``kind_to_fc_parameters`` dictionary that corresponds to the features and parameters (!) from
the tsfresh features that were filtered by the ``some_feature_selection`` feature selection algorithm.


CHANGES.rst
=========
Changelog
=========

tsfresh uses `Semantic Versioning <http://semver.org/>`_

Version 0.20.1
==============

- Added Features
    - Make tsfresh compatible with numpy 1.24 (#1018) and pandas 2.0 (#1028)

- Bugfixes/Typos/Documentation:
    - Use pandas Index.equals in check_if_pandas_series (#963)
    - Updates to package layout, CI/CD and developer setup


Version 0.20.0
==============

- Breaking Change
    - The matrixprofile package becomes an optional dependency

- Bugfixes/Typos/Documentation:
    - Fix feature extraction of Friedrich coefficients for pandas>1.3.5
    - Fix file paths after example notebooks were moved

Version 0.19.0
==============

- Breaking Change
    - Drop Python 3.6 support due to dependency on statsmodels 0.13

- Added Features
    - Improve documentation (#831, #834, #851, #853, #870)
    - Add absolute_maximum and mean_n_absolute_max features (#833)
    - Make settings pickable (#845, #847, #910)
    - Disable multiprocessing for `n_jobs=1` (#852)
    - Add black, isort, and pre-commit (#876)

- Bugfixes/Typos/Documentation:
    - Fix conversion of time-series into sequence for lempel_ziv_complexity (#806)
    - Fix range count config (#827)
    - Reword documentation (#893)
    - Fix statsmodels deprecation issues (#898, #912)
    - Fix typo in requirements (#903)
    - Bump statsmodels to v0.13 (#
    - Updated references


Version 0.18.0
==============

- Added Features
    - Allow arbitrary rolling sizes (#766)
    - Allow for multiclass significance tests (#762)
    - Add multiclass option to RelevantFeatureAugmenter (#782)
    - Addition of matrix_profile feature (#793)
    - Added new query similarity counter feature (#798)
    - Add root mean square feature (#813)
- Bugfixes/Typos/Documentation:
    - Do not send coverage of notebook tests to codecov (#759)
    - Fix typos in notebook (#757, #780)
    - Fix output format of `make_forecasting_frame` (#758)
    - Fix badges and remove benchmark test
    - Fix BY notebook plot (#760)
    - Ts forecast example improvement (#763)
    - Also surpress warnings in dask (#769)
    - Update relevant_feature_augmenter.py (#779)
    - Fix column names in quick_start.rst (#778)
    - Improve relevance table function documentation (#781)
    - Fixed #789 Typo in "how to add custom feature" (#790)
    - Convert to the correct type on warnings (#799)
    - Fix minor typos in the docs (#802)
    - Add unwanted filetypes to gitignore (#819)
    - Fix build and test failures (#815)
    - Fix imputing docu (#800)
    - Bump the scikit-learn version (#822)

Version 0.17.0
==============

We changed the default branch from "master" to "main".

- Breaking Change
    - Changed constructed id in roll_time_series from string to tuple (#700)
    - Same for add_sub_time_series_index (#720)
- Added Features
    - Implemented the Lempel-Ziv-Complexity and the Fourier Entropy (#688)
    - Prevent #524 by adding an assert for common identifiers (#690)
    - Added permutation entropy (#691)
    - Added a logo :-) (#694)
    - Implemented the benford distribution feature (#689)
    - Reworked the notebooks (#701, #704)
    - Speed up the result pivoting (#705)
    - Add a test for the dask bindings (#719)
    - Refactor input data iteration to need less memory (#707)
    - Added benchmark tests (#710)
    - Make dask a possible input format (#736)
- Bugfixes:
    - Fixed a bug in the selection, that caused all regression tasks with un-ordered index to be wrong (#715)
    - Fixed readthedocs (#695, #696)
    - Fix spark and dask after #705 and for non-id named id columns (#712)
    - Fix in the forecasting notebook (#729)
    - Let tsfresh choose the value column if possible (#722)
    - Move from coveralls github action to codecov (#734)
    - Improve speed of data processing (#735)
    - Fix for newer, more strict pandas versions (#737)
    - Fix documentation for feature calculators (#743)

Version 0.16.0
==============

- Breaking Change
    - Fix the sorting of the parameters in the feature names (#656)
      The feature names consist of a sorted list of all parameters now.
      That used to be true for all non-combiner features, and is now also true for combiner features.
      If you relied on the actual feature name, this is a breaking change.
    - Change the id after the rolling (#668)
      Now, the old id of your data is still kept. Additionally, we improved the way
      dataframes without a time column are rolled and how the new sub-time series
      are named.
      Also, the documentation was improved a lot.
- Added Features
    - Added variation coefficient (#654)
    - Added the datetimeindex explanation from the notebook to the docs (#661)
    - Optimize RelevantFeatureAugmenter to avoid re-extraction (#669)
    - Added a function `add_sub_time_series_index` (#666)
    - Added Dockerfile
    - Speed optimizations and speed testing script (#681)
- Bugfixes
    - Increase the extracted `ar` coefficients to the full parameter range. (#662)
    - Documentation fixes (#663, #664, #665)
    - Rewrote the `sample_entropy` feature calculator (#681)
      It is now faster and (hopefully) more correct.
      But your results will change!


Version 0.15.1
==============

- Changelog and documentation fixes

Version 0.15.0
==============

- Added Features
    - Add count_above and count_below feature (#632)
    - Add convenience bindings for dask dataframes and pyspark dataframes (#651)
- Bugfixes
    - Fix documentation build and feature table in sphinx (#637, #631, #627)
    - Add scripts to API documentation
    - Skip dask test for older python versions (#649)
    - Add missing distributor keyword (#648)
    - Fix tuple input for cwt (#645)

Version 0.14.1
==============

- Fix travis deployment

Version 0.14.0
==============

- Breaking Change
    - Replace Benjamini-Hochberg implementation with statsmodels implementation (#570)
- Refactoring and Documentation
    - travis.yml (#605)
    - gitignore (#608)

tsfresh/feature_extraction/settings.py
# -*- coding: utf-8 -*-
# This file as well as the whole tsfresh package are licenced under the MIT licence (see the LICENCE.txt)
# Maximilian Christ (maximilianchrist.com), Blue Yonder Gmbh, 2016
"""
This file contains methods/objects for controlling which features will be extracted when calling extract_features.
For the naming of the features, see :ref:`feature-naming-label`.
"""
import logging
from builtins import range
from collections import UserDict
from inspect import getfullargspec
from itertools import product

import cloudpickle
import pandas as pd

from tsfresh.feature_extraction import feature_calculators
from tsfresh.utilities.string_manipulation import get_config_from_string

_logger = logging.getLogger(__name__)


def from_columns(columns, columns_to_ignore=None):
    """
    Creates a mapping from kind names to fc_parameters objects
    (which are itself mappings from feature calculators to settings)
    to extract only the features contained in the columns.
    To do so, for every feature name in columns this method

    1. split the column name into col, feature, params part
    2. decide which feature we are dealing with (aggregate with/without params or apply)
    3. add it to the new name_to_function dict
    4. set up the params

    :param columns: containing the feature names
    :type columns: list of str
    :param columns_to_ignore: columns which do not contain tsfresh feature names
    :type columns_to_ignore: list of str

    :return: The kind_to_fc_parameters object ready to be used in the extract_features function.
    :rtype: dict
    """

    kind_to_fc_parameters = {}

    if columns_to_ignore is None:
        columns_to_ignore = []

    for col in columns:
        if col in columns_to_ignore:
            continue

        if not isinstance(col, str):
            raise TypeError("Column name {} should be a string or unicode".format(col))

        # Split according to our separator into <col_name>, <feature_name>, <feature_params>
        parts = col.split("__")
        n_parts = len(parts)

        if n_parts == 1:
            raise ValueError(
                "Splitting of columnname {} resulted in only one part.".format(col)
            )

        kind = parts[0]
        feature_name = parts[1]

        if kind not in kind_to_fc_parameters:
            kind_to_fc_parameters[kind] = {}

        if not hasattr(feature_calculators, feature_name):
            raise ValueError("Unknown feature name {}".format(feature_name))

        config = get_config_from_string(parts)
        if config:
            if feature_name in kind_to_fc_parameters[kind]:
                kind_to_fc_parameters[kind][feature_name].append(config)
            else:
                kind_to_fc_parameters[kind][feature_name] = [config]
        else:
            kind_to_fc_parameters[kind][feature_name] = None

    return kind_to_fc_parameters


def include_function(func, exclusion_attr="input_type"):
    """Helper function for selecting specific subset of functions subject to an exclusion attribute
    and the availability of optional dependencies.

    :param func: function to be tested for inclusion
    :type func: object
    :param exclusion_attr: function attribute qualifying as exclusion criterion
    :type exclusion_attr: str

    :return: Boolean indicating if the specific function matches the inclusion criteria.
    :rtype: bool
    """
    decision = (
        hasattr(func, "fctype")
        and not hasattr(func, exclusion_attr)
        and not (
            hasattr(func, "dependency_available")
            and getattr(func, "dependency_available") is False
        )
    )
    return decision


class PickableSettings(UserDict):
    """Base object for all settings, which is a pickable dict.
    For user-specified functions, the settings dictionary might include functions as a key.
    These functions unfortunately can not easily be transported to workers in multiprocessing
    or multi-cloud setups, as they are not pickable by default.
    Therefore, we change the pickle-behavior of this class and use cloudpickle for
    pickling and unpickling the keys of the dictionary, before pickling the full object.
    cloudpickle is able to pickle much more functions than pickle can and pickle will
    only see the already encoded keys (not the raw functions).
    """

    def __getstate__(self):
        """Called on pickling. Encode the keys by cloudpickling them"""
        state = {cloudpickle.dumps(key): value for key, value in self.items()}
        return state

    def __setstate__(self, state):
        """Called on un-pickling. cloudunpickle the keys again"""
        state = {cloudpickle.loads(key): value for key, value in state.items()}
        # please note that the internal dictionary is stored as "data" in the UserDict
        self.__dict__.update(data=state)


# todo: this classes' docstrings are not completely up-to-date
class ComprehensiveFCParameters(PickableSettings):
    """
    Create a new ComprehensiveFCParameters instance. You have to pass this instance to the
    extract_feature instance.

    It is basically a dictionary (and also based on one), which is a mapping from
    string (the same names that are in the feature_calculators.py file) to a list of dictionary of parameters,
    which should be used when the function with this name is called.

    Only those strings (function names), that are keys in this dictionary, will be later used to extract
    features - so whenever you delete a key from this dict, you disable the calculation of this feature.

    You can use the settings object with

    >>> from tsfresh.feature_extraction import extract_features, ComprehensiveFCParameters
    >>> extract_features(df, default_fc_parameters=ComprehensiveFCParameters())

    to extract all features (which is the default nevertheless) or you change the ComprehensiveFCParameters
    object to other types (see below).
    """

    def __init__(self):
        name_to_param = {}

        for name, func in feature_calculators.__dict__.items():
            if (
                callable(func)
                and hasattr(func, "fctype")
                and len(getfullargspec(func).args) == 1
            ):
                name_to_param[name] = None

        name_to_param.update(
            {
                "time_reversal_asymmetry_statistic": [
                    {"lag": lag} for lag in range(1, 4)
                ],
                "c3": [{"lag": lag} for lag in range(1, 4)],
                "cid_ce": [{"normalize": True}, {"normalize": False}],
                "symmetry_looking": [{"r": r * 0.05} for r in range(20)],
                "large_standard_deviation": [{"r": r * 0.05} for r in range(1, 20)],
                "quantile": [
                    {"q": q} for q in [0.1, 0.2, 0.3, 0.4, 0.6, 0.7, 0.8, 0.9]
                ],
                "autocorrelation": [{"lag": lag} for lag in range(10)],
                "agg_autocorrelation": [
                    {"f_agg": s, "maxlag": 40} for s in ["mean", "median", "var"]
                ],
                "partial_autocorrelation": [{"lag": lag} for lag in range(10)],
                "number_cwt_peaks": [{"n": n} for n in [1, 5]],
                "number_peaks": [{"n": n} for n in [1, 3, 5, 10, 50]],
                "binned_entropy": [{"max_bins": max_bins} for max_bins in [10]],
                "index_mass_quantile": [
                    {"q": q} for q in [0.1, 0.2, 0.3, 0.4, 0.6, 0.7, 0.8, 0.9]
                ],
                "cwt_coefficients": [
                    {"widths": width, "coeff": coeff, "w": w}
                    for width in [(2, 5, 10, 20)]
                    for coeff in range(15)
                    for w in (2, 5, 10, 20)
                ],
                "spkt_welch_density": [{"coeff": coeff} for coeff in [2, 5, 8]],
                "ar_coefficient": [
                    {"coeff": coeff, "k": k} for coeff in range(10 + 1) for k in [10]
                ],
                "change_quantiles": [
                    {"ql": ql, "qh": qh, "isabs": b, "f_agg": f}
                    for ql in [0.0, 0.2, 0.4, 0.6, 0.8]
                    for qh in [0.2, 0.4, 0.6, 0.8, 1.0]
                    for b in [False, True]
                    for f in ["mean", "var"]
                    if ql < qh
                ],
                "fft_coefficient": [
                    {"coeff": k, "attr": a}
                    for a, k in product(["real", "imag", "abs", "angle"], range(100))
                ],
                "fft_aggregated": [
                    {"aggtype": s} for s in ["centroid", "variance", "skew", "kurtosis"]
                ],
                "energy_content_frequency_brackets": [
                    {"number_of_bins": 10, "with_normalization": False}
                ],
                "value_count": [{"value": value} for value in [0, 1, -1]],
                "range_count": [
                    {"min": -1, "max": 1},
                    {"min": -1e12, "max": 0},
                    {"min": 0, "max": 1e12},
                ],
                "approximate_entropy": [
                    {"m": 2, "r": r} for r in [0.1, 0.3, 0.5, 0.7, 0.9]
                ],
                "friedrich_coefficients": (
                    lambda m: [
                        {"coeff": coeff, "m": m, "r": 30} for coeff in range(m + 1)
                    ]
                )(3),
                "max_langevin_fixed_point": [{"m": 3, "r": 30}],
                "linear_trend": [
                    {"attr": "pvalue"},
                    {"attr": "rvalue"},
                    {"attr": "intercept"},
                    {"attr": "slope"},
                    {"attr": "stderr"},
                ],
                "agg_linear_trend": [
                    {"attr": attr, "chunk_len": i, "f_agg": f}
                    for attr in ["rvalue", "intercept", "slope", "stderr"]
                    for i in [5, 10, 50]
                    for f in ["max", "min", "mean", "var"]
                ],
                "augmented_dickey_fuller": [
                    {"attr": "teststat"},
                    {"attr": "pvalue"},
                    {"attr": "usedlag"},
                ],
                "number_crossing_m": [{"m": 0}, {"m": -1}, {"m": 1}],
                "energy_ratio_by_chunks": [
                    {"num_segments": 10, "segment_focus": i} for i in range(10)
                ],
                "ratio_beyond_r_sigma": [
                    {"r": x} for x in [0.5, 1, 1.5, 2, 2.5, 3, 5, 6, 7, 10]
                ],
                "linear_trend_timewise": [
                    {"attr": "pvalue"},
                    {"attr": "rvalue"},
                    {"attr": "intercept"},
                    {"attr": "slope"},
                    {"attr": "stderr"},
                ],
                "count_above": [{"t": 0}],
                "count_below": [{"t": 0}],
                "lempel_ziv_complexity": [{"bins": x} for x in [2, 3, 5, 10, 100]],
                "fourier_entropy": [{"bins": x} for x in [2, 3, 5, 10, 100]],
                "permutation_entropy": [
                    {"tau": 1, "dimension": x} for x in [3, 4, 5, 6, 7]
                ],
                "query_similarity_count": [{"query": None, "threshold": 0.0}],
                "matrix_profile": [
                    {"threshold": 0.98, "feature": f}
                    for f in ["min", "max", "mean", "median", "25", "75"]
                ],
                "mean_n_absolute_max": [
                    {
                        "number_of_maxima": 3,
                        "number_of_maxima": 5,
                        "number_of_maxima": 7,
                    }
                ],
            }
        )

        # remove missing dependencies
        for name, func in feature_calculators.__dict__.items():
            if (
                callable(func)
                and hasattr(func, "dependency_available")
                and getattr(func, "dependency_available") is False
            ):
                name_to_param.pop(name)
                _logger.warning(
                    f"Dependency not available for {name}, this feature will be disabled!"
                )

        super().__init__(name_to_param)


class MinimalFCParameters(ComprehensiveFCParameters):
    """
    This class is a child class of the ComprehensiveFCParameters class
    and has the same functionality as its base class. The only difference is,
    that most of the feature calculators are disabled and only a small
    subset of calculators will be calculated at all. Those are denoted by an attribute called "minimal".

    Use this class for quick tests of your setup before calculating all
    features which could take some time depending of your data set size.

    You should use this object when calling the extract function, like so:

    >>> from tsfresh.feature_extraction import extract_features, MinimalFCParameters
    >>> extract_features(df, default_fc_parameters=MinimalFCParameters())
    """

    def __init__(self):
        ComprehensiveFCParameters.__init__(self)

        for fname, f in feature_calculators.__dict__.items():
            if fname in self and (
                not hasattr(f, "minimal") or not getattr(f, "minimal")
            ):
                del self[fname]


class EfficientFCParameters(ComprehensiveFCParameters):
    """
    This class is a child class of the ComprehensiveFCParameters class
    and has the same functionality as its base class.

    The only difference is, that the features with high computational costs are not calculated. Those are denoted by
    the attribute "high_comp_cost".

    You should use this object when calling the extract function, like so:

    >>> from tsfresh.feature_extraction import extract_features, EfficientFCParameters
    >>> extract_features(df, default_fc_parameters=EfficientFCParameters())
    """

    def __init__(self):
        ComprehensiveFCParameters.__init__(self)

        # drop all features with high computational costs
        for fname, f in feature_calculators.__dict__.items():
            if fname in self and hasattr(f, "high_comp_cost"):
                del self[fname]


class IndexBasedFCParameters(ComprehensiveFCParameters):
    """
    This class is a child class of the ComprehensiveFCParameters class
    and has the same functionality as its base class.

    The only difference is that only the features that require a pd.Series as an input are
    included. Those have an attribute "input" with value "pd.Series".
    """

    def __init__(self):
        ComprehensiveFCParameters.__init__(self)
        # drop all features with high computational costs
        for fname, f in feature_calculators.__dict__.items():
            if fname in self and getattr(f, "input", None) != "pd.Series":
                del self[fname]


class TimeBasedFCParameters(ComprehensiveFCParameters):
    """
    This class is a child class of the ComprehensiveFCParameters class
    and has the same functionality as its base class.

    The only difference is, that only the features that require a DatetimeIndex are included. Those
    have an attribute "index_type" with value pd.DatetimeIndex.
    """

    def __init__(self):
        ComprehensiveFCParameters.__init__(self)
        # drop all features with high computational costs
        for fname, f in feature_calculators.__dict__.items():
            if fname in self and getattr(f, "index_type", False) != pd.DatetimeIndex:
                del self[fname]
